app/src/main/java/com/example/splendormobilegame/activities/GameActivity/BuyingReservedCardsController.java

```diff
@@ -44,15 +61,39 @@ public class BuyReservedCardMessageHandler extends UserReaction {
         public UserMessage react(ServerMessage serverMessage) {
             Log.i("UserReaction", "Entered BuyReservedCardMessageHandler react method");
 
-            // TODO Update the model
-            // TODO Update the view via `gameActivity` or other objects given in constructor
-
+            //Get the data from the request
+            BuyReservedMine.ResponseData responseData = (BuyReservedMine.ResponseData) ReactionUtils.getResponseData(
+                    serverMessage,BuyReservedMine.ResponseData.class
+            );
+
+            BuyReservedMine.TokensDataResponse tokensDataResponse = responseData.buyer.tokens;
+            EnumMap<TokenType,Integer> tokens = new EnumMap<TokenType,Integer>(TokenType.class);
+            tokens.put(TokenType.RUBY,tokensDataResponse.ruby);
+            tokens.put(TokenType.EMERALD,tokensDataResponse.emerald);
+            tokens.put(TokenType.SAPPHIRE,tokensDataResponse.sapphire);
+            tokens.put(TokenType.DIAMOND,tokensDataResponse.diamond);
+            tokens.put(TokenType.ONYX,tokensDataResponse.onyx);
+            tokens.put(TokenType.GOLD_JOKER,tokensDataResponse.gold);
+
+            // Update the model
+            Room room = Model.getRoom();
+            Game game = room.getGame();
+            User buyer = room.getUserByUuid(responseData.buyer.userUuid);
+            Card boughtCard = game.getCardByUuid(responseData.buyer.cardUuid);
+            ReservedCard tempReservedCard = new ReservedCard(boughtCard, buyer, false);
+
+
+            buyer.addCard(boughtCard);
+            game.removeReservedCard(buyer, tempReservedCard);
```

Comment:
You may want to create new method in the `Game` like `removeReservedCard(UUID userUuid, UUID cardUuid)`. This method will search the reservedCards for a match and remove it then.

Reply:
I've made it your way 👍